src/legacy/core_plugins/kibana/public/visualize/np_ready/editor/lib/visualize_app_state.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { createHashHistory } from 'history';
+import { isFunction, omit } from 'lodash';
+
+import { migrateAppState } from './migrate_app_state';
+import {
+  createKbnUrlStateStorage,
+  createStateContainer,
+  syncState,
+} from '../../../../../../../../plugins/kibana_utils/public';
+import { PureVisState, VisualizeAppState, VisualizeAppStateTransitions } from '../../types';
+
+const STATE_STORAGE_KEY = '_a';
+
+interface Arguments {
+  useHash: boolean;
+  stateDefaults: VisualizeAppState;
+}
+
+function toObject(state: PureVisState): PureVisState {
+  return omit(state, (value, key: string) => {
+    return key.charAt(0) === '$' || key.charAt(0) === '_' || isFunction(value);
+  });
+}
+
+export function useVisualizeAppState({ useHash, stateDefaults }: Arguments) {
+  const history = createHashHistory();
+  const kbnUrlStateStorage = createKbnUrlStateStorage({
```

Comment:
Just for info: to fix pinned/unpinned filter, it will be important that global state syncing uses the same kbnUrlStateStorage as app state syncing.